ast.c

```diff
@@ -144,33 +150,21 @@ ast_s_parse_file(rb_execution_context_t *ec, VALUE module, VALUE path, VALUE kee
 static VALUE
 rb_ast_parse_file(VALUE path, VALUE keep_script_lines, VALUE error_tolerant, VALUE keep_tokens)
 {
-    VALUE f;
     VALUE ast_value = Qnil;
+    VALUE f = rb_file_open_str(path, "r");
     rb_encoding *enc = rb_utf8_encoding();
 
-    f = rb_file_open_str(path, "r");
     rb_funcall(f, rb_intern("set_encoding"), 2, rb_enc_from_encoding(enc), rb_str_new_cstr("-"));
-    VALUE vparser = ast_parse_new();
-    if (RTEST(keep_script_lines)) rb_parser_set_script_lines(vparser);
-    if (RTEST(error_tolerant))  rb_parser_error_tolerant(vparser);
-    if (RTEST(keep_tokens))  rb_parser_keep_tokens(vparser);
-    ast_value = rb_parser_compile_file_path(vparser, Qnil, f, 1);
+    ast_value = rb_ast_compile(keep_script_lines, error_tolerant, keep_tokens, f, rb_parser_compile_file_path);
     rb_io_close(f);
```

Comment:
When `rb_parser_compile_file_path` failed, `f` will be left open.
I think that it is too much to pass a function pointer to the new compile function.
What about a function that creates a parser then just sets the options?